### ISSUE TITLE
BUG: Quiche will deadlock when packet got lost

### ISSUE THREAD

**Issue by zylthinking** [cloudflare/quiche#2254]
Hello, 
I met a dead lock bug, which in fact, happens very easily.

I will post some log and analyse it after that:

The log shows:
```
0000000000000000000000000000000100000000 rx frm ACK delay=0 blocks=[40290..41079, 41081..41085, 41089..41442, 41444..42066, 42068..42070, 42072..42072, 42074..42085, 42088..42088, 42090..42137, 42139..42178, 42180..42182, 42184..42184, 42186..42308, 42310..42869, 42871..42877, 42879..42887, 42889..42889, 42891..42891, 42895..43324, 43326..43594, 43596..43602, 43605..43613, 43616..43622, 43625..43625, 43627..43628, 43630..43630, 43638..43638, 43641..44254, 44256..44290, 44292..44293, 44295..44296, 44304..44304, 44308..44719, 44721..44898, 44900..44906, 44908..44914, 44918..44920, 44922..44922, 44930..44931, 44933..44934, 44936..44941, 44944..44951, 44954..44954, 44958..45931, 45933..46382, 46384..46385, 46388..46388, 46391..46391, 46393..46394, 46398..46402, 46404..46404, 46409..46410, 46414..46415, 46417..46418, 46420..46875, 46877..46878, 46880..46883, 46886..46892, 46903..47141, 47143..47168, 47170..47175, 47188..47313, 47315..47326, 47332..47929, 47931..48342] ecn_counts=None
0000000000000000000000000000000100000000 rx frm ACK delay=0 blocks=[40290..41079, 41081..41085, 41089..41442, 41444..42066, 42068..42070, 42072..42072, 42074..42085, 42088..42088, 42090..42137, 42139..42178, 42180..42182, 42184..42184, 42186..42308, 42310..42869, 42871..42877, 42879..42887, 42889..42889, 42891..42891, 42895..43324, 43326..43594, 43596..43602, 43605..43613, 43616..43622, 43625..43625, 43627..43628, 43630..43630, 43638..43638, 43641..44254, 44256..44290, 44292..44293, 44295..44296, 44304..44304, 44308..44719, 44721..44898, 44900..44906, 44908..44914, 44918..44920, 44922..44922, 44930..44931, 44933..44934, 44936..44941, 44944..44951, 44954..44954, 44958..45931, 45933..46382, 46384..46385, 46388..46388, 46391..46391, 46393..46394, 46398..46402, 46404..46404, 46409..46410, 46414..46415, 46417..46418, 46420..46875, 46877..46878, 46880..46883, 46886..46892, 46903..47141, 47143..47168, 47170..47175, 47188..47313, 47315..47326, 47332..47929, 47931..48343] ecn_counts=None
will sleep Some(4.999840265s)
check peer window Ok(0)
will sleep None

check peer window Ok(0) after recv Ok((436, 65.49.223.94:443))  
check peer window Ok(0) after get something
will sleep None

check peer window Ok(0) after recv Ok((436, 65.49.223.94:443))
check peer window Ok(0) after get something

check peer window Ok(0) after recv Ok((1060, 65.49.223.94:443))
check peer window Ok(0) after get something
will sleep None
```

Based on the logs, my suspected process might be as follows:
1. A packet sent by one peer is lost and therefore needs to be retransmitted.
2. Before the retransmitted packet arrives, other packets continue to arrive, eventually exceeding the flow control window and causing it to close.
3. The retransmitted packet never arrives.
4. In this situation, even receiving an ACK from the peer or opportunistically calling on_timeout is useless. on_timeout does not do anything.
5. The peer eventually receives all packets it can (up to the closed window), so it won't send any more ACKs. The peer is now stuck in a perpetual recv wait.
6. The local peer is also waiting for the other side to open its window, entering a perpetual wait.

you can checking the tcpdump output corresponse the log:
```
18:34:56.161601 IP 65.49.223.94.16clouds.com.443 > linux.49796: UDP, length 193  ------ the acks but 40289 is lost
18:34:56.161604 IP 65.49.223.94.16clouds.com.443 > linux.49796: UDP, length 193
18:34:56.161605 IP 65.49.223.94.16clouds.com.443 > linux.49796: UDP, length 193
18:34:56.163047 IP 65.49.223.94.16clouds.com.443 > linux.49796: UDP, length 193
18:34:56.176478 IP 65.49.223.94.16clouds.com.443 > linux.49796: UDP, length 193
18:34:56.176485 IP 65.49.223.94.16clouds.com.443 > linux.49796: UDP, length 193
18:34:56.176485 IP 65.49.223.94.16clouds.com.443 > linux.49796: UDP, length 193
18:34:56.176546 IP 65.49.223.94.16clouds.com.443 > linux.49796: UDP, length 193
18:34:56.176546 IP 65.49.223.94.16clouds.com.443 > linux.49796: UDP, length 193
18:34:56.176547 IP 65.49.223.94.16clouds.com.443 > linux.49796: UDP, length 193
18:34:56.176547 IP 65.49.223.94.16clouds.com.443 > linux.49796: UDP, length 193
18:34:56.176547 IP 65.49.223.94.16clouds.com.443 > linux.49796: UDP, length 193

----------- at here, the control window closed: which reported by the log above:
check peer window Ok(0)
will sleep None

18:35:21.321677 IP 65.49.223.94.16clouds.com.443 > linux.49796: UDP, length 436
check peer window Ok(0) after recv Ok((436, 65.49.223.94:443))  ---- checking socket recv_from returns
check peer window Ok(0) after get something  ------------------ checking after quiche conn.recv() returns and after trying flushing response conn.recv() gernerated (nothing send)
will sleep None   -------- select! { recv; sleep until conn.timeout which is None }

18:36:25.831959 IP 65.49.223.94.16clouds.com.443 > linux.49796: UDP, length 436
18:36:25.831966 IP 65.49.223.94.16clouds.com.443 > linux.49796: UDP, length 1060
```

**Comment by zylthinking** [comment]
believe this is the key point: The mechanism allows in-order packets to be treated as acknowledged (acked) even when the preceding packet has not been acked. This removes them from the Congestion Window (CWND), allowing data exceeding the Sliding Window size to be sent despite the CWND being smaller than the Sliding Window. Consequently, the Sliding Window gets filled up.

```
            for unacked in self.sent_packets.range_mut(start..) {
                if unacked.pkt_num >= peer_sent_range.end {
                    break;
                }

                if unacked.time_acked.is_some() {
                    // Already acked.
                } else if unacked.time_lost.is_some() {
                    // An acked packet was already declared lost.
                    spurious_losses += 1;
                    spurious_pkt_thresh
                        .get_or_insert(largest_acked - unacked.pkt_num + 1);
                    unacked.time_acked = Some(now);

                    if unacked.in_flight {
                        has_in_flight_spurious_loss = true;
                    }
                } else {
                    if unacked.in_flight {
                        self.in_flight_count -= 1;
                        acked_bytes += unacked.size;
                    }

                    newly_acked.push(Acked {
                        pkt_num: unacked.pkt_num,
                        time_sent: unacked.time_sent,
                        size: unacked.size,

                        rtt: now.saturating_duration_since(unacked.time_sent),
                        delivered: unacked.delivered,
                        delivered_time: unacked.delivered_time,
                        first_sent_time: unacked.first_sent_time,
                        is_app_limited: unacked.is_app_limited,
                    });

                    trace!("{} packet newly acked {}", trace_id, unacked.pkt_num);

                    self.acked_frames
                        .extend(std::mem::take(&mut unacked.frames));

                    has_ack_eliciting |= unacked.ack_eliciting;
                    unacked.time_acked = Some(now);
                }
            }
```

**Comment by zylthinking** [comment]
This bug should be the sending part believes it has recieved an ack for a packet by wrong.
Because I can see the `loss timer cleared` in the log which means self.bytes_in_flight.is_zero() 
While, the other side get Done from stream_recv, I think the reason is there are packets lost in the stream.

```rust
    fn set_loss_detection_timer(&mut self, handshake_status: HandshakeStatus, now: Instant) {
        if let (Some(earliest_loss_time), _) = self.loss_time_and_space() {
            // println!("loss timer set");
            // Time threshold loss detection.
            self.loss_timer.update(earliest_loss_time);
            return;
        }

        if self.bytes_in_flight.is_zero() && handshake_status.peer_verified_address {
            // println!("loss timer cleared");
            self.loss_timer.clear();
            return;
        }

        // PTO timer.
        if let (Some(timeout), _) = self.pto_time_and_space(handshake_status, now) {
            // println!("loss timer set");
            self.loss_timer.update(timeout);
        }
    }

```

This function returns false, which be called by stream_recv --> steam.is_readable() and I am sure there are data in the stream.
So, the reason should be packet lost, but self.bytes_in_flight.is_zero()  does not agree with it
```rust
    pub fn ready(&self) -> bool {
        let (_, buf) = match self.data.first_key_value() {
            Some(v) => v,
            None => return false,
        };

        buf.off() == self.off
    }
```
